elliptic-curve/src/hash2field.rs

```diff
@@ -2,32 +2,32 @@ mod expand_msg;
 mod expand_msg_xmd;
 mod expand_msg_xof;
 
-use core::convert::TryFrom;
 pub use expand_msg::*;
 pub use expand_msg_xmd::*;
 pub use expand_msg_xof::*;
+use generic_array::{typenum::Unsigned, ArrayLength, GenericArray};
 
 /// The trait for helping to convert to a scalar
-pub trait FromOkm<const L: usize>: Sized {
+pub trait FromOkm {
+    /// The number of bytes needed to convert to a scalar
+    type Length: ArrayLength<u8>;
+
     /// Convert a byte sequence into a scalar
-    fn from_okm(data: &[u8; L]) -> Self;
+    fn from_okm(data: &GenericArray<u8, Self::Length>) -> Self;
 }
 
 /// Convert an arbitrary byte sequence according to
 /// <https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-11#section-5.3>
-pub fn hash_to_field<E, T, const L: usize, const COUNT: usize, const OUT: usize>(
-    data: &[u8],
-    domain: &[u8],
-) -> [T; COUNT]
+pub fn hash_to_field<E, T>(data: &[u8], domain: &'static [u8], out: &mut [T])
```

Comment:
I think it would be great to add a `hash_to_scalar` method to `GroupDigest`. For example the OPAQUE protocol requires the `hash_to_field` function to derive a `Scalar`.

Reply:
Can you give an example of your preference for `hash_to_scalar`?

Reply:
I believe this is already it. `hash_to_scalar` would just be used to build a `Scalar` directly from `out`.
See https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-13.html#section-5-4:
> The hash_to_field function is also suitable for securely hashing to scalars. For example, when hashing to the scalar field for an elliptic curve (sub)group with prime order r, it suffices to instantiate hash_to_field with target field GF(r).

I'm unsure what "instantiate hash_to_field with target field GF(r)" means exactly. Please correct me if I'm just talking nonsense here.

Reply:
Actually thinking about this, my initial goal was to provide `HashToScalar` from the VOPRF spec: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-voprf-08#section-4.1. But reading now through this again, VOPRF specifies specific routines for every group that aren't really part of `hash_to_scalar` from the hash2curve spec.

So I don't need this anymore. But maybe it's still useful for other people?

Reply:
I believe hash-to-scalar should already be covered by the `Reduce` trait: https://docs.rs/elliptic-curve/latest/elliptic_curve/ops/index.html

See `k256` as an example (for some reason this renders badly in rustdoc): https://docs.rs/k256/latest/src/k256/arithmetic/scalar.rs.html#586-590

You can also use `Reduce` with `NonZeroScalar` if you would like the result to be guaranteed to be non-zero: this performs the same wide reduction on a `U512`, but reduced modulo MODULUS - 1, and then adds one.

Unfortunately this is only implemented for `k256` at present.

Reply:
I actually just discovered the `NonZeroScalar` part. That's amazing, and it's infallible too!

Reply:
Here's a PR to add `digest` support to the `Reduce` trait: https://github.com/RustCrypto/traits/pull/869/files